### PR TITLE
MATLAB multiphase OCP solver

### PR DESCRIPTION
**Pull request by FreyJo**
Simulink is not supported yet.

### CODE REVIEW

File: interfaces/acados_template/acados_template/c_templates_tera/matlab_templates/acados_mex_set.in.c [hunk 1]
```diff
@@ -550,36 +588,47 @@ void mexFunction(int nlhs, mxArray *plhs[], int nrhs, const mxArray *prhs[])
...
+    {%- endfor %}
+    {%- set np_max = np_values | sort | last %}
+{% else %}
+    {%- set np_max = dims.np %}
+{%- endif %}
+
+{% if np_max > 0 %}
         MEX_DIM_CHECK_MAT(fun_name, field, nrow, ncol, nrow, 2);
         // create int index vector
         // TODO: MOCP get max np!
```
Comment (sandmaennchen) [inline interfaces/acados_template/acados_template/c_templates_tera/matlab_templates/acados_mex_set.in.c:612]: remove comment ?

Comment (sandmaennchen) [inline interfaces/acados_template/acados_template/c_templates_tera/matlab_templates/acados_mex_set.in.c:612]: remove comment ?

File: interfaces/acados_matlab_octave/AcadosMultiphaseOptions.m [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    methods
+        function obj = AcadosMultiphaseOptions()
+            obj.integrator_type = {};
+            obj.collocation_type = {};
+            obj.cost_discretization = {};
+        end
+
+        function make_consistent(self, opts, n_phases)
+            % check if all fields are lists of length n_phases
+            INTEGRATOR_TYPES = {'ERK', 'IRK', 'GNSF'};
```
Comment (sandmaennchen) [inline interfaces/acados_matlab_octave/AcadosMultiphaseOptions.m:47]: LIFTED_IRK? :D

File: examples/acados_matlab_octave/mocp_transition_example/main_multiphase_ocp.m [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+
+settings = get_example_settings();
+
+N_list = [10, 1, 15];
+n_phases = length(N_list);
+N_horizon = sum(N_list);
+
+% create_multiphase_ocp_solver
+ocp = AcadosMultiphaseOcp(N_list);
```
Comment (sandmaennchen) [inline examples/acados_matlab_octave/mocp_transition_example/main_multiphase_ocp.m:38]: Maybe call this `mocp` and rename `phase_1` to `ocp_phase_1` etc. to emphasize that it is just three regular OCP object.

File: interfaces/acados_matlab_octave/AcadosMultiphaseOcp.m [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+                error('phase_idx must be less than or equal to the number of phases');
+            end
+
+            % Check solver options
+            non_default_opts = find_non_default_fields_of_obj(ocp.solver_options);
+
+            if ~isempty(non_default_opts)
+                fprintf('WARNING: set_phase: Phase %d contains non-default solver options: %s, which will be ignored.\n', ...
+                        phase_idx, strjoin(non_default_opts, ', '));
+                fprintf('Solver options need to be set via AcadosMultiphaseOcp.solver_options or mocp_opts instead.\n');
```
Comment (sandmaennchen) [inline interfaces/acados_matlab_octave/AcadosMultiphaseOcp.m:124]: maybe better: 'The solver option %s needs to be set via ...' to make it more explicit that not all options are in the MOCP Options.

Comment (FreyJo) [reply interfaces/acados_matlab_octave/AcadosMultiphaseOcp.m]: I improved the message.
However this is just a warning, if the phase OCP is set up in a way that also options are set, I want to only print one warning for all options, as done now.
It is the same in python.

File: examples/acados_matlab_octave/mocp_transition_example/main_multiphase_ocp.m (hunk 3 above)
Comment (FreyJo) [reply examples/acados_matlab_octave/mocp_transition_example/main_multiphase_ocp.m:38]: I sticked to the naming in the corresponding Python example, which is also in the tutorial example in our paper, so I would keep it the same here.
We might change it when revising the paper.

File: interfaces/acados_matlab_octave/find_non_default_fields_of_obj.m [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            default_val = dummy_obj_struct.(field);
+
+            if ~isequal(val, default_val)
+                nondefault_fields{end+1} = field;
+            end
+        end
+    end
+end
+
+function result = endsWith(strArray, suffix)
```
Comment (sandmaennchen) [inline interfaces/acados_matlab_octave/find_non_default_fields_of_obj.m:87]: could be implemented with `reverse()` and `strncmp` ..

Comment (FreyJo) [reply interfaces/acados_matlab_octave/find_non_default_fields_of_obj.m:87]: The 'reverse' function is not yet implemented in Octave.

Comment (sandmaennchen) [reply interfaces/acados_matlab_octave/find_non_default_fields_of_obj.m:87]: :cry:

**Pull request merged**